### ISSUE TITLE
Does reapidjson support converting an instance of a struct into json string?

### ISSUE THREAD

**Issue by GingerMoon** [Tencent/rapidjson#1299]
Does reapidjson support converting an instance of a struct into json string?
And if yes, can the order of the fields be kept in the order of the declaration in the struct?
For example,
struct S {
int Z;
int A;
};
S s = {1,2};

I tried to read the doc but not able to find the answer.
The expect json string is:
{"Z":1, "A":2}

Any sample code would be much appreciated.

**Comment by miloyip** [comment]
No. RapidJSON does not support a framework for automatic serialization/deserialization. Some libraries based on RapidJSON can do so, e.g. [Cereal](https://github.com/USCiLab/cereal).

However, you may also write code for serialization of your types, e.g.:
~~~cpp
void Serialize(Writer<StringBuffer>& writer, const S& s) {
    writer.StartObject();
    writer.Key("Z");
    writer.Int(s.Z);
    writer.Key("A");
    writer.Int(s.A);
    writer.EndObject();
}

void foo() {
    S s = {1, 2};
    StringBuffer sb;
    Writer<StringBuffer> writer(sb);
    Serialize(writer, s);
    std::cout << sb.GetString() << std::endl;
}
~~~

For more information, check http://rapidjson.org/md_doc_sax.html#Writer

**Comment by GingerMoon** [comment]
Thanks a lot @miloyip!
I will spend more time on Cereal for the serialization, can I ask you a question, may be stupid, is there any method like writer.Object(...) so that the embedded struct  in S can also be printed?
I checked the doc  http://rapidjson.org/md_doc_sax.html#Writer but not able to find one.

**Comment by miloyip** [comment]
No. `Writer` only helps to generate a JSON from those 7 JSON types, not custom defined types.
So you cannot automatically serialize any struct, instead you need to manually write down how to map a struct to JSON.
